src/impl.jl

```diff
@@ -0,0 +1,337 @@
+using Pkg
+
+const is_expr = Base.Meta.isexpr
+
+function pkgversion(m::Module)
+    pkgdir = dirname(string(first(methods(m.eval)).file))
+    toml = Pkg.TOML.parsefile(joinpath(pkgdir, "..", "Project.toml"))
+    VersionNumber(toml["version"])
+end
+
+function if_has_package(
+    action::Function,
+    name::String,
+    uuid::Base.UUID,
+    version::VersionNumber
+)
+    pkgid = Base.PkgId(uuid, name)
+    if Base.root_module_exists(pkgid)
+        pkg = Base.root_module(pkgid)
+        if pkgversion(pkg) >= version
+            return action(pkg)
+        end
+    end
+end
+
+function error_usage(__source__)
+    usage=
```

Comment:
Is there a way to get the doc string rather than repeating it here?

Reply:
I think there is a way to get the docstring, yeah..? I'm not sure what it is now though..

But i think the more common pattern is to make this string into a constant, and then _use it_ both here and in the docstring.

Reply:
Aha, it's with `@doc`:
```julia
julia> docstr = @doc AutoHashEqualsCached.@auto_hash_equals;

julia> typeof(docstr)
Markdown.MD

julia> docstr
  @auto_hash_equals struct Foo ... end

  Produces specializations of Base.hash and Base.==:

    •  Base.== is implemented as an elementwise test for isequal.

    •  Base.hash combines the elementwise hash code of the fields with the hash code of the type's simple name.

  The hash code and == implementations ignore type parameters, so that Box{Int}(1) will be considered isequal to Box{Any}(1). This is for
  compatibility with the package AutoHashEquals.jl.
```

Which is incidentally also the way to _set_ a docstring to a value:
```julia
julia> function foo end
foo (generic function with 0 methods)

julia> @doc docstr foo
foo

help?> foo
search: foo floor pointer_from_objref OverflowError RoundFromZero unsafe_copyto! functionloc StackOverflowError @functionloc OutOfMemoryError

  @auto_hash_equals struct Foo ... end

  Produces specializations of Base.hash and Base.==:

    •  Base.== is implemented as an elementwise test for isequal.

    •  Base.hash combines the elementwise hash code of the fields with the hash code of the type's simple name.

  The hash code and == implementations ignore type parameters, so that Box{Int}(1) will be considered isequal to Box{Any}(1). This is for
  compatibility with the package AutoHashEquals.jl.
```
So you could set the docstring with the constant this way, or fetch it this way. Up to you.